chain/client/src/stateless_validation/state_witness_producer.rs

```diff
@@ -163,88 +163,89 @@ impl Client {
         let mut implicit_transitions = vec![];
-        for block_hash in implicit_blocks {
-            let state_transition = store
-                .get_ser(
-                    near_store::DBCol::StateTransitionData,
-                    &near_primitives::utils::get_block_shard_id(block_hash, shard_id),
-                )?
-                .ok_or_else(|| {
-                    let message = format!(
-                        "Missing implicit transition state proof for block {block_hash} and shard {shard_id}"
-                    );
-                    if !cfg!(feature = "shadow_chunk_validation") {
-                        log_assert_fail!("{message}");
-                    }
-                    Error::Other(message)
-                })?;
-            match state_transition {
-                StoredChunkStateTransitionData::V1(StoredChunkStateTransitionDataV1 {
-                    base_state,
-                    ..
-                }) => {
-                    implicit_transitions.push(ChunkStateTransition {
-                        block_hash: *block_hash,
-                        base_state,
-                        post_state_root: *self
-                            .chain
-                            .get_chunk_extra(block_hash, &shard_uid)?
-                            .state_root(),
-                    });
-                }
+
+        loop {
+            let header = self.chain.get_block_header(&current_block_hash)?;
+            if header.height() < prev_chunk_height_included {
+                return Err(Error::InvalidBlockHeight(prev_chunk_height_included));
+            }
+
+            let current_epoch_id = *header.epoch_id();
+            let current_shard_id =
+                self.epoch_manager.get_prev_shard_id(&current_block_hash, next_shard_id)?.0;
+            if current_shard_id != next_shard_id {
+                // If shard id changes, we need to get implicit state
+                // transition from current shard id to the next shard id.
+                let (chunk_state_transition, _, _) =
+                    self.get_state_transition(&current_block_hash, &next_epoch_id, next_shard_id)?;
+                implicit_transitions.push(chunk_state_transition);
+            }
+            next_epoch_id = current_epoch_id;
+            next_shard_id = current_shard_id;
+
+            if header.height() == prev_chunk_height_included {
+                break;
             }
+
+            // Add implicit state transition.
+            let (chunk_state_transition, _, _) = self.get_state_transition(
+                &current_block_hash,
+                &current_epoch_id,
+                current_shard_id,
+            )?;
+            implicit_transitions.push(chunk_state_transition);
+
+            current_block_hash = *header.prev_hash();
         }
 
+        let main_block = current_block_hash;
+        let epoch_id = next_epoch_id;
+        let shard_id = next_shard_id;
+        implicit_transitions.reverse();
```

Comment:
Ok so resharding, if it exits, is always the first state transition to be applied

Reply:
Not really. The first transition is always the main one, then implicit transitions come, and if there is resharding transition in B, it will come after usual implicit transition in B.